src/vm/index.md

```diff
@@ -94,20 +95,42 @@ Then the following registers are initialized (without explicit initialization, a
 
 ## Contexts
 
-There are 3 _contexts_ in the FuelVM: [predicates](#predicate-verification), [scripts](#script-execution), and [calls](./instruction_set.md#call-call-contract). A context is an isolated execution environment with defined [memory ownership](#ownership) and can be _external_ or _internal_:
+There are 4 _contexts_ in the FuelVM: [predicate estimation](#predicate-estimation), [predicate verification](#predicate-verification), [scripts](#script-execution), and [calls](./instruction_set.md#call-call-contract). A context is an isolated execution environment with defined [memory ownership](#ownership) and can be _external_ or _internal_:
 
 - External: predicate and script. `$fp` will be zero.
 - Internal: call. `$fp` will be non-zero.
 
 [Returning](./instruction_set.md#return-return-from-call) from a context behaves differently depending on whether the context is external or internal.
 
+## Predicate Estimation
+
+For any input of type [`InputType.Coin`](../protocol/tx_format/index.md), a non-zero `predicateLength` field means the UTXO being spent is a [P2SH](https://en.bitcoinwiki.org/wiki/P2SH) rather than a [P2PKH](https://en.bitcoinwiki.org/wiki/Pay-to-Pubkey_Hash) output.
+
+For each such input in the transaction, the VM is [initialized](#vm-initialization), then:
+
+1. `$pc` and `$is` are set to the start of the input's `predicate` field.
+1. `$ggas` and `$cgas` are set to `tx.gasLimit`.
+
+Predicate verification will fail if gas is exhausted during execution.
+
+After successful execution, `predicate.gasUsed` is set to `tx.gasLimit - $cgas`.
```

Comment:
Even though we don't allow calls, I think it's safer to use ggas instead of cgas.

```suggestion
After successful execution, `predicate.gasUsed` is set to `tx.gasLimit - $ggas`.
```